### ISSUE TITLE
Fatal error: memory exhausted

### ISSUE THREAD

**Issue by svivian** [NoahY/q2a-badges#1]
Just installed the badge plugin. I activated the plugin, then selected all badges and saved. Then I clicked "Recheck Badge Awarding" but get a fatal error:

**Fatal error: Allowed memory size of 134217728 bytes exhausted (tried to allocate 97 bytes) in /qapath/qa-include/qa-db.php on line 530**

I'm guessing it's because I have a pretty huge database right now: nearly 3k users, 7k questions, over 14k answers and 15k comments. I tried ticking only a couple of badges but still get the same error.

Might be worth trying an AJAX style recalculation, like Q2A does with user points. Run a separate AJAX query for each few users or something.

**Comment by NoahY** [comment]
Dear Scott,

Okay, that's probably a reason why Gideon has his rebuilding functions 
go in stages via ajax... I guess that's what will have to happen with 
the Badges plugin as well... really, it would be nice if badges were in 
the core so they could take advantage of that ajax routine - I tried 
integrating with the ajax rebuild in the first place, but got stuck 
somewhere, can't remember where.

Until then, maybe you could try upping the php ram allowance 
temporarily?  Not sure how much ram it actually needs for 7k questions :)

Best,

Noah

**Comment by NoahY** [comment]
Okay, I've made the function unset each variable as it uses it... this 
seems to impact the memory usage considerably on testing, though you'll 
have to see what it does for CPU usage on large tables.  The manual page 
for unset seems to suggest it's not a bad way to retrieve memory.

Let me know if this helps.

**Comment by svivian** [comment]
Thanks, will test later and let you know. By the way, 134217728 bytes is 128 MB which is already _much higher_ than PHP's default memory allocation of 16 MB. Usually large memory increases like that come from loading many complex objects at once, for example loading an entire database table into memory. Even if the table is only a few MB the memory used in PHP will be several times that amount.

BTW if you didn't already know, you can call memory_get_peak_usage() towards the end of a script to get the maximum memory used in that script. Useful for debugging.

**Comment by NoahY** [comment]
Nice, I was using get_memory_usage() or whatever it is called... nice to 
have a peak usage.  I'll check and see if I can tweak that.  Let me know 
how the new version works out... I'll see if I can save more memory 
somehow, otherwise I'll work on an ajax version that breaks it up into 
parts.

I know, 128 is pretty high.  It's crazy that this little script could 
take more than that; maybe I'm really doing something I shouldn't.

**Comment by Arech** [comment]
Hi, friends.

I've just installed badges plugin from master branch and experiencing the same problem after it activation when trying to open /admin/plugins page. Further more, sometimes PHP process crashes with "memory size 128Mb exhausted" error and some-times it just crashes silently without any errors at all (this time php5-fpm.log says something like WARNING: [pool www] child <number_here> exited on signal 11 (SIGSEGV)).

Tried to set memory_limit to 256mb or 180mb later, but it didn't help. The only difference is that PHP started to crash with SIGSEGV everytime.

I'm using nginx/1.2.6 and PHP 5.3.20-1~dotdeb.0 with Suhosin-Patch (cli) (built: Dec 24 2012 11:46:09) plugged in FPM mode.

I'll try to debug it and I'll write later if I'll find something... Any guess where to start from?

**Comment by Arech** [comment]
Found infinite recursion. The problem is around function qa-plugin.php::qa_badge_name() and it first statement - call to qa_opt(). qa_opt() causes qa_badge_admin::option_default() to be called, which in turn cause a call to qa_badge_name() within case 'badge_'.$slug.'_name'.

Don't know yet how to fix it... Give me your ideas, please.

**Comment by Arech** [comment]
It looks like the root cause is wrong call to qa_badge_name() in option_default().
Check first case clause within a switch. Call to qa_badge_name('badges/'.$slug); uses different notation than it does usually. I guess, it should be qa_lang('badges/'.$slug); like this:

``` php
switch($option) {
    case 'badge_'.$slug.'_name':
        //return qa_badge_name('badges/'.$slug);
        return qa_lang('badges/'.$slug);
```

It fixes the recursion and works (at least as far as I can see from now).

I'll open a new bug report for this buq and move all my comments there. @svivian sorry to bother, I thought the bug was similar...

**Comment by NoahY** [comment]
should be fixed now.
